jupyterhub_traefik_proxy/etcd.py

```diff
@@ -59,52 +75,47 @@ def _default_client(self):
         help="""The etcd key prefix for traefik dynamic configuration""",
     )
 
-    async def _setup_traefik_static_config(self):
-        await super()._setup_traefik_static_config()
-        kv = {}
+    @run_on_executor
+    def _etcd_transaction(self, success_actions):
+        status, response = self.etcd_client.transaction(
+            compare=[], success=success_actions, failure=[]
+        )
+        return status, response
 
-        def get_etcd_kvs(d, etcd_key):
-            for k, v in d.items():
-                if isinstance(v, dict):
-                    new_key = etcd_key + k + "/"
-                    get_etcd_kvs(v, new_key)
-                else:
-                    kv[etcd_key + k] = v
+    @run_on_executor
+    def _etcd_get(self, key):
+        value, _ = self.etcd_client.get(key)
+        return value
 
-        get_etcd_kvs(self.static_config, self.etcd_traefik_prefix)
+    @run_on_executor
+    def _etcd_get_prefix(self, prefix):
+        routes = self.etcd_client.get_prefix(prefix)
+        return routes
 
-        success = [
-            KV.put.txn(
-                self.etcd_traefik_prefix + "etcd/endpoint",
-                str(urlparse(self.etcd_url).hostname)
-                + ":"
-                + str(urlparse(self.etcd_url).port),
-            ),
-            KV.put.txn(
-                self.etcd_traefik_prefix + "etcd/prefix", self.etcd_traefik_prefix
-            ),
-            KV.put.txn(self.etcd_traefik_prefix + "etcd/useapiv3", "true"),
-            KV.put.txn(self.etcd_traefik_prefix + "etcd/watch", "true"),
-            KV.put.txn(self.etcd_traefik_prefix + "providersThrottleDuration", "1"),
-        ]
+    async def _setup_traefik_static_config(self):
+        await super()._setup_traefik_static_config()
 
-        for key, value in kv.items():
-            if isinstance(value, bool):
-                value = str(value).lower()
-            if isinstance(value, list):
-                for i in range(len(value)):
-                    key += "/" + str(i)
-                    value = value[i]
-            success.append(KV.put.txn(key.lower(), value))
-
-        status, response = await self.etcd_client.txn(
-            compare=[], success=success, fail=[]
-        )
+        self.static_config["etcd"] = {
+            "username": self.etcd_username,
+            "password": self.etcd_password,
+            "endpoint": str(urlparse(self.etcd_url).hostname)
+            + ":"
+            + str(urlparse(self.etcd_url).port),
+            "prefix": self.etcd_traefik_prefix,
+            "useapiv3": True,
+            "watch": True,
+            "providersThrottleDuration": 1,
+        }
 
-        if status:
-            self.log.error(
-                "Couldn't set up traefik's static config. Response: %s", response
+        try:
+            traefik_utils.persist_static_conf(
+                self.toml_static_config_file, self.static_config
             )
+        except IOError:
```

Comment:
Let's always raise here. As a general rule, we only want to proceed past an error if it's recoverable. Failing to write this file should mean a hard exit, so letting it raise every time is probably what we want.

Reply:
You're right! I've just added this to the toml version also.